cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/ControllerFactory.java

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.runtime.distributed.remote;
+
+import com.google.common.util.concurrent.Service;
+import io.cdap.cdap.app.runtime.ProgramOptions;
+import io.cdap.cdap.common.lang.ClassLoaders;
+import io.cdap.cdap.common.logging.LoggingContext;
+import io.cdap.cdap.common.logging.LoggingContextAccessor;
+import io.cdap.cdap.common.ssh.SSHConfig;
+import io.cdap.cdap.internal.app.runtime.ProgramOptionConstants;
+import io.cdap.cdap.internal.app.runtime.SystemArguments;
+import io.cdap.cdap.internal.tethering.runtime.spi.provisioner.TetheringProvisioner;
+import io.cdap.cdap.logging.context.LoggingContextHelper;
+import io.cdap.cdap.proto.id.ProgramRunId;
+import io.cdap.cdap.runtime.spi.RuntimeMonitorType;
+import io.cdap.cdap.runtime.spi.runtimejob.RuntimeJobManager;
+import org.apache.twill.api.TwillController;
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.common.Threads;
+import org.apache.twill.internal.ServiceListenerAdapter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.Nullable;
+
+/**
+ * TwillControllerFactory implementation that is used by {@link RemoteExecutionTwillRunnerService}.
+ */
+class ControllerFactory implements TwillControllerFactory {
```

Comment:
I'm assuming this is just the existing code being pulled out into a separate class. Please comment on any part that is actually different, and also see other comment on whether we really need to move it out.

Reply:
Moved ControllerFactory back to the RemoteExecutionTwillRunnerService.